### DISCUSSION TITLE
Support for Error Events and Loading Progress

### DISCUSSION THREAD

**Discussion by itzzritik** [katspaugh/wavesurfer.js#3055] [Q&A] [answered]
I was on Wavesurfer 6 that was having `loading` and `error` events
which was helpful in tracking load progress and error events.

After migrating to version 7, Looks like they are removed,
How to show loading progress or catch load errors now?

Any Example code would be appreciated.

**Comment by katspaugh** [comment]
In v7, there's currently only a `load` event (fires when loading starts) and `ready` when it's done.

You can catch errors directly on the load method like this:
```js
try {
  await wavesurfer.load(url)
} catch (e) {
  console.log('Wavesurfer loading failed', e)
}
```

I think it would be good to restore a loading progress event. It's doable via the ReadableSteam API. I'll make an issue for this.

**Reply by rsxdalv** [reply]
Tried it with WaveSurfer.create and it is unable to catch the errors.

```ts
try {
      const ws = WaveSurfer.create({
        ...options,
        container: containerRef.current,
      });
      setWavesurfer(ws);
      return () => {
        ws.destroy();
      };
    } catch (e) {
      console.log("=== Wavesurfer loading failed", e);
    }
```
I still get uncaught (in Promise) errors.

**Reply by rsxdalv** [reply]
Had to workaround like so:
```ts

    const ws = WaveSurfer.create({
      ...options,
      url: undefined,
      container: containerRef.current,
    });
    ws.load(options.url!).catch((e) => {
      console.log("=== Wavesurfer loading failed", e);
    });
```

**Comment by thijstriemstra** [comment]
> I was on Wavesurfer 6 that was having loading and error events
which was helpful in tracking load progress and error events.

They're essential. 

> it would be good to restore a loading progress event

Why the would you remove it in the first place?

**Comment by katspaugh** [comment, marked as answer]
I didn't initially include it in v7 because one of its goals was improved performance.
Tracking the progress of a fetch request requires [reading the downloaded data chunks](https://github.com/katspaugh/wavesurfer.js/pull/3063/files#diff-eb275dd515bd68f45e5e14b12d3f4ff97dc9b1e52d742f0e66f56f67dccb21b0R9-R30) which is not good for performance compared to just reading it into a Blob, and might even cause memory leaks.
